### PR TITLE
feat: gwas catalog top-hit + study step

### PR DESCRIPTION
**Pull request by d0choa**
Step to run the top-hits in isolation of everything else.

The `GWASCatalogTopHitIngestionStep` included here was run in dataproc in 17 minutes.

Subsequent PRs will handle GWAS Catalog + Summary Statistics + PICS road which might result in removing some steps.

### CODE REVIEW

File: src/gentropy/datasource/gwas_catalog/study_index.py [hunk 1]
```diff
@@ -332,7 +330,6 @@ def from_source(
         return (
             cls._parse_study_table(catalog_studies)
             .annotate_ancestries(ancestry_file)
-            .annotate_sumstats_info(sumstats_lut)
```
Comment (d0choa) [inline src/gentropy/datasource/gwas_catalog/study_index.py:335]: This change can be problematic. Not on top-hits, or sumstats processing but for the generation of the curation table. @DSuveges we can discussed but I think @addramir was suggesting to not care too much about curation for now

**Comment by d0choa** [comment]
I'm having second thoughts. Because the 2 DAGs have interdependencies in the business logic, I want to see both working before we review this.

I'm reverting to draft

**Comment by d0choa** [comment]
I'll summarise the status of this PR, and I'd appreciate some thoughts @DSuveges and @project-defiant 

In general it isn't straightforward because it touches four gwas catalog processes:
a) top hits
b) sumstats pics
c) susie finemapping
d) gwas catalog curation table

With the current state of the PR, the DAGs for a) and b) run to completion without considering curation or sumstats QC c) is not yet implemented, but I made a pass to the logic to generate the study index in the context of QC sumstats and/or availability of sumstats and/or curation, but it hasn't been tested.  d) I didn't pay too much attention to it.

In terms of features. This PR contains:

- Addition of the `gwas_catalog_study_index` step, which is a refactor of the previous `gwas_catalog_study_inclusion`. Now, this step does not aim to filter; just generates a study index as complete as possible, allowing for 3 optional parameters: sum stats QC, OT curation and availability of summary statistics (ideally from our harmonised bucket). This step could be reused in b-d but never in a) which contains its own study index using the study splitting.

- A new `gwas_catalog_top_hits` step, which simplifies the older `gwas_catalog_ingestion` step. This step produces `StudyLocus` from GWAS catalog associations and their respective StudyIndex. It no longer tries to consolidate information on summary statistics, significantly simplifying the process.

- A new class to parse the OT curation, by just moving the existing code to a separate class.

It needs to be better tested. I might be distracted for a few days, but I'm happy to return and add more unit tests unless this work becomes essential for something else earlier.

**Comment by d0choa** [comment]
## Example how to run the step:

```bash
poetry run gentropy step=gwas_catalog_study_index \
step.catalog_study_files="[ '/Users/ochoa/Datasets/gwas_catalog_download_studies.tsv' ]" \
step.catalog_ancestry_files="[ '/Users/ochoa/Datasets/gwas_catalog_download_ancestries.tsv' ]" \
step.study_index_path=/Users/ochoa/Datasets/study_index_annotated \
step.gwas_catalog_study_curation_file=/Users/ochoa/Datasets/20241004_output_curation.tsv \
step.sumstats_qc_path=/Users/ochoa/Datasets/20241015_GwasCatQCLogs
```

## Copy of the outputs

```
gs://ot-team/dochoa/study_index_annotated
```

## Breakdown of GWAS Catalog flags

```
+----------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----+
|studyType |analysisFlags          |qualityControls                                                                                                                                                                                                                                    |count|
+----------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----+
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., Harmonized summary statistics are not available or empty.]                                                                                                                              |54168|
|gwas      |[]                     |[]                                                                                                                                                                                                                                                 |16842|
|pQTL      |[]                     |[]                                                                                                                                                                                                                                                 |16533|
|gwas      |[Metabolite]           |[]                                                                                                                                                                                                                                                 |12064|
|gwas      |[ExWAS]                |[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |3695 |
|gwas      |[Metabolite]           |[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |2849 |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets.]                                                                                                                                                                                         |2145 |
|Microbiome|[]                     |[]                                                                                                                                                                                                                                                 |1128 |
|gwas      |[Metabolite]           |[The number of SNPs in the study is below the expected threshold.]                                                                                                                                                                                 |1033 |
|gwas      |[]                     |[The mean beta QC check value is not within the expected range.]                                                                                                                                                                                   |793  |
|gwas      |[]                     |[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |604  |
|Microbiome|[]                     |[The mean beta QC check value is not within the expected range.]                                                                                                                                                                                   |414  |
|gwas      |[ExWAS]                |[The number of SNPs in the study is below the expected threshold.]                                                                                                                                                                                 |294  |
|gwas      |[]                     |[The PZ QC check values are not within the expected range.]                                                                                                                                                                                        |267  |
|gwas      |[]                     |[The number of SNPs in the study is below the expected threshold.]                                                                                                                                                                                 |255  |
|gwas      |[Multivariate analysis]|[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |248  |
|pQTL      |[]                     |[The number of SNPs in the study is below the expected threshold.]                                                                                                                                                                                 |200  |
|Microbiome|[]                     |[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |138  |
|gwas      |[GxE]                  |[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |73   |
|gwas      |[Case-case study]      |[]                                                                                                                                                                                                                                                 |68   |
|gwas      |[GxE]                  |[]                                                                                                                                                                                                                                                 |51   |
|gwas      |[Non-additive model]   |[]                                                                                                                                                                                                                                                 |45   |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The number of SNPs in the study is below the expected threshold.]                                                                                                                       |41   |
|gwas      |[GxG]                  |[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |29   |
|gwas      |[]                     |[The GC lambda value is not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                                                                                          |27   |
|gwas      |[Non-additive model]   |[The mean beta QC check value is not within the expected range., The PZ QC check values are not within the expected range., The GC lambda value is not within the expected range.]                                                                 |25   |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The PZ QC check values are not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                            |21   |
|gwas      |[GxG]                  |[]                                                                                                                                                                                                                                                 |21   |
|Microbiome|[]                     |[The PZ QC check values are not within the expected range.]                                                                                                                                                                                        |20   |
|gwas      |[]                     |[The mean beta QC check value is not within the expected range., The PZ QC check values are not within the expected range.]                                                                                                                        |16   |
|gwas      |[GxE]                  |[The number of SNPs in the study is below the expected threshold.]                                                                                                                                                                                 |16   |
|gwas      |[Multivariate analysis]|[]                                                                                                                                                                                                                                                 |15   |
|gwas      |[]                     |[The mean beta QC check value is not within the expected range., The GC lambda value is not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                          |11   |
|gwas      |[Non-additive model]   |[The PZ QC check values are not within the expected range.]                                                                                                                                                                                        |11   |
|gwas      |[]                     |[The mean beta QC check value is not within the expected range., The PZ QC check values are not within the expected range., The GC lambda value is not within the expected range.]                                                                 |10   |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The PZ QC check values are not within the expected range.]                                                                                                                              |9    |
|gwas      |[]                     |[The PZ QC check values are not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                                                                                      |9    |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The mean beta QC check value is not within the expected range., The PZ QC check values are not within the expected range., The GC lambda value is not within the expected range.]       |7    |
|gwas      |[]                     |[The GC lambda value is not within the expected range.]                                                                                                                                                                                            |7    |
|gwas      |[Case-case study]      |[Harmonized summary statistics are not available or empty.]                                                                                                                                                                                        |7    |
|gwas      |[Non-additive model]   |[The mean beta QC check value is not within the expected range., The PZ QC check values are not within the expected range.]                                                                                                                        |6    |
|gwas      |[]                     |[The mean beta QC check value is not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                                                                                 |6    |
|gwas      |[]                     |[The mean beta QC check value is not within the expected range., The GC lambda value is not within the expected range.]                                                                                                                            |5    |
|pQTL      |[]                     |[The PZ QC check values are not within the expected range.]                                                                                                                                                                                        |3    |
|gwas      |[Metabolite]           |[The PZ QC check values are not within the expected range.]                                                                                                                                                                                        |3    |
|Microbiome|[]                     |[The PZ QC check values are not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                                                                                      |3    |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The GC lambda value is not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                                |2    |
|gwas      |[ExWAS]                |[]                                                                                                                                                                                                                                                 |2    |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The mean beta QC check value is not within the expected range., The PZ QC check values are not within the expected range.]                                                              |2    |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The mean beta QC check value is not within the expected range.]                                                                                                                         |2    |
|gwas      |[Case-case study]      |[The PZ QC check values are not within the expected range.]                                                                                                                                                                                        |2    |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The mean beta QC check value is not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                       |2    |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The PZ QC check values are not within the expected range., The GC lambda value is not within the expected range.]                                                                       |1    |
|gwas      |[]                     |[The PZ QC check values are not within the expected range., The GC lambda value is not within the expected range.]                                                                                                                                 |1    |
|gwas      |[]                     |[The mean beta QC check value is not within the expected range., The PZ QC check values are not within the expected range., The number of SNPs in the study is below the expected threshold.]                                                      |1    |
|gwas      |[Non-additive model]   |[The PZ QC check values are not within the expected range., The GC lambda value is not within the expected range.]                                                                                                                                 |1    |
|gwas      |[]                     |[GWAS Catalog study has not been curated by Open Targets., The mean beta QC check value is not within the expected range., The GC lambda value is not within the expected range., The number of SNPs in the study is below the expected threshold.]|1    |
+----------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----+
```

File: src/gentropy/datasource/gwas_catalog/associations.py [hunk 2]
```diff
@@ -1108,7 +1106,10 @@ def from_source(
         pvalue_threshold is keeped in sync with the WindowBasedClumpingStep gwas_significance.
         """
         return StudyLocusGWASCatalog(
-            _df=gwas_associations.withColumn(
+            _df=gwas_associations
+            # drop duplicate rows
+            .distinct()
+            .withColumn(
                 "studyLocusId", f.monotonically_increasing_id().cast(StringType())
```
[the remark this answers is not in the document]
Comment (d0choa) [reply src/gentropy/datasource/gwas_catalog/associations.py:1113]: I asked the same to @DSuveges . It's a temporary ID that doesn't make it to the final dataset

File: src/gentropy/gwas_catalog_top_hits.py [hunk 3]
```diff
@@ -10,83 +10,55 @@
...
+            variant_annotation_path (str): Path to GnomAD variants.
             catalog_studies_out (str): Output GWAS catalog studies path.
             catalog_associations_out (str): Output GWAS catalog associations path.
             distance (int): Distance, within which tagging variants are collected around the semi-index.
-            gwas_catalog_study_curation_file (str | None): file of the curation table. Optional.
-            inclusion_list_path (str | None): optional inclusion list (parquet)
         """
         # Extract
-        gnomad_variants = VariantIndex.from_parquet(session, gnomad_variant_path)
+        gnomad_variants = VariantIndex.from_parquet(session, variant_annotation_path)
```
[the remark this answers is not in the document]
Comment (d0choa) [reply src/gentropy/gwas_catalog_top_hits.py:44]: In theory, it should be the gnomAD variant annotation dataset in the schema of a variantIndex. We haven't run it for a while so I wonder if everything works fine there after the ETL VariantIndex work. This DAG should have no dependency on the ETL VariantIndex

**Comment by addramir** [comment]
Added changes in ancestry mapping in [gwas_population_2_LD_panel_map.json](https://github.com/opentargets/gentropy/pull/808/files#diff-8a572375d8f148e3b6e5706f1a376b1c8fcbac54d44c0bee792f16343024e880) since we are going to remake the study index. These changes add CSA (central south asians) from UKBB to the list of ancestries.

File: src/gentropy/assets/data/gwas_population_2_LD_panel_map.json
Comment (d0choa) [inline src/gentropy/assets/data/gwas_population_2_LD_panel_map.json:1]: @addramir are you sure this won't break PICS if LDIndex does not contain `csa`?
Can you please double-check?

Comment (addramir) [reply src/gentropy/assets/data/gwas_population_2_LD_panel_map.json:1]: What will happen with PICS if the major ancestry code form the study (e.g. csa) is not in LD index for PICS @DSuveges @vivienho ?

Comment (vivienho) [reply src/gentropy/assets/data/gwas_population_2_LD_panel_map.json:1]: @addramir In that case R will be 0 and the variant should be dropped before PICS.

Comment (addramir) [reply src/gentropy/assets/data/gwas_population_2_LD_panel_map.json:1]: In this case we need to revert the change with changing mapping. I will add a quick temporary fix in susie but in the future we have to change that.

Comment (addramir) [reply src/gentropy/assets/data/gwas_population_2_LD_panel_map.json:1]: Reverted

Comment (project-defiant) [reply src/gentropy/assets/data/gwas_population_2_LD_panel_map.json:1]: @addramir @d0choa can I run the test from the branch to clump the sumstats for susie already?

**Review by DSuveges — APPROVED** [review]
I'm approving the PR assuming there might be some surprises along the way when it comes to curation.

**Pull request merged**